ruma-serde/src/canonical_json/value.rs

```diff
@@ -0,0 +1,244 @@
+    /// # use serde_json::json;
+    /// # use std::convert::TryInto;
+    /// # use ruma_serde::CanonicalJsonValue;
+    /// let v: CanonicalJsonValue = json!(["an", "array"]).try_into().unwrap();
+    /// ```
+    Array(Vec<CanonicalJsonValue>),
+
+    /// Represents a JSON object.
+    ///
+    /// The map is backed by a BTreeMap to guarantee the sorting of keys.
+    ///
+    /// ```
+    /// # use serde_json::json;
+    /// # use std::convert::TryInto;
+    /// # use ruma_serde::CanonicalJsonValue;
+    /// let v: CanonicalJsonValue = json!({ "an": "object" }).try_into().unwrap();
+    /// ```
+    Object(BTreeMap<String, CanonicalJsonValue>),
+}
+
+impl CanonicalJsonValue {
+    /// Returns a canonical JSON string according to Matrix specification.
+    ///
+    /// The method should be preferred over `serde_json::to_string` since it
+    /// checks the size of the canonical string. Matrix canonical JSON enforces
+    /// a size limit of less than 65,535 when sending PDU's for the server-server protocol.
+    pub fn to_canonical_string(&self) -> Result<String, serde_json::Error> {
+        Ok(serde_json::to_string(self).and_then(|s| {
+            if s.len() > 65_535 {
+                Err(serde_json::Error::custom("TOO LARGE"))
+            } else {
+                Ok(s)
+            }
+        })?)
+    }
+}
+
+impl fmt::Debug for CanonicalJsonValue {
+    fn fmt(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+        match *self {
+            CanonicalJsonValue::Null => formatter.debug_tuple("Null").finish(),
+            CanonicalJsonValue::Bool(v) => formatter.debug_tuple("Bool").field(&v).finish(),
+            CanonicalJsonValue::Integer(ref v) => fmt::Debug::fmt(v, formatter),
+            CanonicalJsonValue::String(ref v) => formatter.debug_tuple("String").field(v).finish(),
+            CanonicalJsonValue::Array(ref v) => {
+                formatter.write_str("Array(")?;
+                fmt::Debug::fmt(v, formatter)?;
+                formatter.write_str(")")
+            }
+            CanonicalJsonValue::Object(ref v) => {
+                formatter.write_str("Object(")?;
+                fmt::Debug::fmt(v, formatter)?;
+                formatter.write_str(")")
+            }
+        }
+    }
+}
+
+impl fmt::Display for CanonicalJsonValue {
+    /// Display a JSON value as a string.
+    ///
+    /// ```
+    /// # use serde_json::json;
+    /// #
+    /// let json = json!({ "city": "London", "street": "10 Downing Street" });
+    ///
+    /// // Canonical format:
+    /// //
+    /// // {"city":"London","street":"10 Downing Street"}
+    /// let compact = format!("{}", json);
+    /// assert_eq!(compact,
+    ///     "{\"city\":\"London\",\"street\":\"10 Downing Street\"}");
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{}", serde_json::to_string(&self).map_err(|_| fmt::Error)?)
+    }
+}
+
+impl TryFrom<JsonValue> for CanonicalJsonValue {
+    type Error = serde_json::Error;
```

Comment:
Why `serde_json::Error`? I think a dedicated error type would be better. The only error condition is having a number with decimal places, or outside the `js_int::Int` range, right?

Reply:
I made an enum with a few variants, hope that works.